Client/mods/deathmatch/logic/lua/CLuaArguments.cpp

```diff
@@ -535,46 +535,51 @@ json_object* CLuaArguments::WriteTableToJSONObject(bool bSerialize, CFastHashMap
 
-    bool                                  bIsArray = true;
-    unsigned int                          iArrayPos = 1;            // lua arrays are 1 based
-    vector<CLuaArgument*>::const_iterator iter = m_Arguments.begin();
+    bool                                      bIsArray = true;
+    vector<pair<unsigned int, CLuaArgument*>> vecSortedArguments;            // lua arrays are not necessarily sorted
+    vector<CLuaArgument*>::const_iterator     iter = m_Arguments.begin();
     for (; iter != m_Arguments.end(); iter += 2)
     {
         CLuaArgument* pArgument = *iter;
         if (pArgument->GetType() == LUA_TNUMBER)
         {
             double       num = pArgument->GetNumber();
             unsigned int iNum = static_cast<unsigned int>(num);
-            if (num == iNum)
-            {
-                if (iArrayPos != iNum)            // check if the value matches its index in the table
-                {
-                    bIsArray = false;
-                    break;
-                }
-            }
-            else
-            {
-                bIsArray = false;
-                break;
-            }
+
+            vecSortedArguments.push_back({iNum, *(iter + 1)});
         }
         else
         {
             bIsArray = false;
             break;
         }
-        iArrayPos++;
     }
 
-    if (bIsArray)
+    if (bIsArray)            // the table could possibly be an array
     {
-        json_object*                          my_array = json_object_new_array();
-        vector<CLuaArgument*>::const_iterator iter = m_Arguments.begin();
-        for (; iter != m_Arguments.end(); iter++)
+        // sort the table based on the keys (already handled correctly by std::pair)
+        std::sort(vecSortedArguments.begin(), vecSortedArguments.end());
+
+        // only the first and last element are checked, everything else is correct by default because the vector was sorted
+        unsigned int const iFirstKey = vecSortedArguments.front().first;
+        unsigned int const iLastKey = vecSortedArguments.back().first;
+
+        unsigned int const iFirstArrayPos = 1;            // lua arrays are 1 based
+        unsigned int const iLastArrayPos = static_cast<unsigned int>(vecSortedArguments.size());
+
+        if (vecSortedArguments.empty() || iFirstKey != iFirstArrayPos || iLastKey != iLastArrayPos)
         {
-            iter++;            // skip the key values
-            CLuaArgument* pArgument = *iter;
+            bIsArray = vecSortedArguments.empty();            // an empty table is also considered an array
+        }
+    }
+
+    if (bIsArray)            // the table is definitely an array
+    {
+        json_object*                                              my_array = json_object_new_array();
+        vector<pair<unsigned int, CLuaArgument*>>::const_iterator iter = vecSortedArguments.begin();
+        for (; iter != vecSortedArguments.end(); ++iter)
```

Comment:
Use for-range loop instead

Reply:
You're right, I was unsure which syntax was preferred in the mta codebase